creator-node/src/snapbackSM.js

```diff
@@ -47,63 +61,387 @@ class SnapbackSM {
+    this.log(`Discovery provider: ${currentlySelectedDiscProv}`)
+    return resp.data.data
+  }
+
+  // Retrieve the current clock value on this node for the provided user wallet
+  async getUserPrimaryClockValues (wallets) {
+    const cnodeUsers = await models.CNodeUser.findAll({
+      where: {
+        walletPublicKey: {
+          [models.Sequelize.Op.in]: wallets
+        }
+      }
+    })
+    return cnodeUsers.reduce((o, k) => {
+      o[k.walletPublicKey] = k.clock
+      return o
+    }, {})
+  }
+
+  // Enqueue a sync request to a particular secondary
+  async issueSecondarySync (userWallet, secondaryEndpoint, primaryEndpoint) {
+    let syncRequestParameters = {
+      baseURL: secondaryEndpoint,
+      url: '/sync',
+      method: 'post',
+      data: {
+        wallet: [userWallet],
+        creator_node_endpoint: primaryEndpoint,
+        state_machine: true // state machine specific flag
+      }
+    }
+    await this.syncQueue.add({ syncRequestParameters, startTime: Date.now() })
+  }
 
+  // Main state machine processing function
+  async processStateMachineOperation (job) {
+    this.log(`------------------Process SnapbackSM Operation, slice ${this.currentModuloSlice}------------------`)
+    if (this.audiusLibs == null) {
+      logger.error(`Invalid libs instance`)
+      return
+    }
+    // Attempt to initialize, if this creator node has not yet been registered on chain
+    // then no operations will be performed
     if (!this.initialized) {
-      this.log(`Initializing SnapbackSM`)
-      this.log(`Retrieving spID for ${endpoint}`)
-      const recoveredSpID = await this.audiusLibs.ethContracts.ServiceProviderFactoryClient.getServiceProviderIdFromEndpoint(
-        endpoint
-      )
-      // A returned spID of 0 means this endpoint is currently not registered on chain
-      // In this case, the stateMachine is considered to be 'uninitialized'
-      if (recoveredSpID === 0) return
-      config.set('spID', recoveredSpID)
-      this.initialized = true
+      await this.initializeNodeIdentityConfig()
+      return
+    }
+    // 1.) Retrieve base information for state machine operations
+    let spInfo = await this.getSPInfo()
+    if (spInfo.spID === 0) {
+      this.log(`Invalid spID, recovering ${spInfo}`)
+      await this.recoverSpID()
+      return
     }
-    this.log(`Recovered ${config.get('spID')} for ${endpoint}`)
+    // Retrieve users list for this node
+    let usersList = await this.getNodePrimaryUsers()
 
-    // TODO: Enable after dev
-    // Run the task every x time interval
-    // this.stateMachineQueue.add({}, { repeat: { cron: '0 */x * * *' } })
+    // Generate list of wallets to query clock number
+    // Structured as { nodeEndpoint: [wallet1, wallet2, ...] }
+    let nodeVectorClockQueryList = {}
+
+    // Users actually selected to process
+    let usersToProcess = []
+
+    let wallets = []
+
+    // Issue queries to secondaries for each user
+    usersList.map(
```

Comment:
super nit but you shouldn't use map for functions where you don't need to return something, instead forEach may be better. also the linter may yell at you for this. try this in the chrome console 
```js
users.map(user => {
    if (user%2 === 0) return
    console.log(user)
})
```
vs 
```
users.forEach(user => {
    if (user%2 === 0) return
    console.log(user)
})
```

Reply:
fix the comment above this